docs/architecture/adr-015-ibc-packet-receiver.md

```diff
@@ -0,0 +1,205 @@
+# ADR 015: IBC Packet Receiver
+
+## Changelog
+
+- 2019 Oct 22: Initial Draft
+
+## Context
+
+[ICS 26 - Routing Module](https://github.com/cosmos/ics/tree/master/spec/ics-026-routing-module) defines function [`handlePacketRecv`](https://github.com/cosmos/ics/tree/master/spec/ics-026-routing-module#packet-relay).
+
+`handlePacketRecv` executes per-module `onRecvPacket` callbacks, verifies the
+packet merkle proof, and pushes the acknowledgement bytes, if present, to the IBC
+channel `Keeper` state (ICS04).
+
+`handlePacketAcknowledgement` executes per-module `onAcknowledgementPacket`
+callbacks, and verifies the acknowledgement commitment proof.
+
+`handlePacketTimeout` and `handlePacketTimeoutOnClose` executes per-module
+`onTimeoutPacket` callbacks, and verifies the timeout proof.
+
+The mechanism is similar to the transaction handling logic in `baseapp`. After
+authentication, the handler is executed, and the authentication state change
+must be committed regardless of the result of the handler execution.
+
+`handlePacketRecv` also requires acknowledgement writing which has to be done
+after the handler execution and also must be commited regardless of the result of
+the handler execution.
+
+## Decision
+
+The Cosmos SDK will define an `AnteDecorator` for IBC packet receiving. The
+`AnteDecorator` will iterate over the messages included in the transaction, type
+`switch` to check whether the message contains an incoming IBC packet, and if so
+verify the Merkle proof.
+
+```go
+// Pseudocode
+type ProofVerificationDecorator struct {
+  clientKeeper ClientKeeper
+  channelKeeper ChannelKeeper
+}
+
+func (pvr ProofVerificationDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+  var flag bool
+
+  for _, msg := range tx.GetMsgs() {
+    var err error
+    switch msg := msg.(type) {
+    case client.MsgUpdateClient:
+      err = pvr.clientKeeper.UpdateClient(msg.ClientID, msg.Header)
+    case channel.MsgPacket:
+      err = pvr.channelKeeper.VerifyPacket(msg.Packet, msg.Proofs, msg.ProofHeight)
+      flag = true
+      // Store the empty acknowledgement for convinience
+      pvr.channelKeeper.SetPacketAcknowledgement(ctx, msg.PortID, msg.ChannelID, msg.Sequence, []byte{})
+    case chanel.MsgAcknowledgement:
+      err = pvr.channelKeeper.VerifyAcknowledgement(msg.Acknowledgement, msg.Proof, msg.ProofHeight)
+      flag = true
+    case channel.MsgTimeoutPacket:
+      err = pvr.channelKeeper.VerifyTimeout(msg.Packet, msg.Proof, msg.ProofHeight, msg.NextSequenceRecv)
+      flag = true
+    default:
+      if flag {
+        return ctx, errors.New("Transaction cannot include both IBC packet messasges and normal messages")
```

Comment:
why is this a problem?

Reply:
An attacker can include a failing message inside the transaction, which will abort the transaction so the packet messages that come after will be ignored but the sequence will be kept increased, which makes any actor able to skip arbitrary packets. Any transaction that includes IBC messages should not be aborted.

Reply:
Hmm, if they do this we should revert the state changes from IBC packets (including sequence increases) as well - it should be possible to have both kinds of messages.

Reply:
is there any potential room for improvement based on @cwgoes's comment?

Reply:
^^

Reply:
The `AnteHandler` state change(which includes sequence increasing) is committed regardless of the app handler result, so we cannot revert the state change after the process entered into the app handler.

Reply:
Ah, hmm, that is a bit of an abstraction mismatch I suppose. 

Could we still allow regular messages & IBC packets as long as they aren't interleaved? (all regular messages first)

Reply:
I still think we should try to find a way to do this. Maybe we can discuss?

Reply:
The optimal solution will be having an after-execution cleanup logic which will increase the sequence only after the packets were successfully executed(where the AnteHandler only verifies the proofs)

Reply:
Should we do that, then?

Reply:
Just to say that I proposed such an architecture for decorators. That they wrapped the handler, which would allow rolling back to a certain point if the handler failed.

This approach was rejected as needless complexity, but I think that decision could be revisited. Indeed, if you want to add after execution cleanup, please don't hack it in, but update the architecture to allow this to be cleanly.done.

Reply:
In this case we want to increase the sequence regardless and instead write a "failure" acknowledgement if the packet processing throws (e.g. if the recipient account doesn't exist or something) - but we plan to use wrapper functions for writing acknowledgements in this fashion, essentially like you are proposing I think.